### ISSUE TITLE
Muxes can get caught in infinite loop.

### ISSUE THREAD

**Issue by ejhumphrey** [pescadores/pescador#87]
In updating another project from pescador==0.1.3 to 1.0, I was running into an issue where a data stream was hanging indefinitely. 

Fortunately, having identified #86 (attempts to re-iterate a mux returned no samples), I was able to figure out why a mux-of-muxes would get caught in an infinite loop. 

**tl:dr;** If a Mux finds itself in a situation where all of its streams are fundamentally exhausted (cannot produce an iterator that yields samples), attempts to iterate will run forever. 

This is documented in a [test case here](https://github.com/pescadores/pescador/blob/ejh_20170403_test_refactor/tests/test_mux.py#L307). 

I *think* [a check here](https://github.com/pescadores/pescador/blob/ejh_20170403_test_refactor/pescador/mux.py#L209) that `distribution.any() > 0` would do the trick... @bmcfee any thoughts?

**Comment by bmcfee** [comment]
> @bmcfee any thoughts?

That does indeed seem like it should fix the issue.  I would have thought that it would trigger a stop iteration once the distribution hits 0 within next_stream, but maybe the logic got messed up somewhere along the refactor.

**Comment by ejhumphrey** [comment]
Working through this now, so partially documenting my thought process, but our fix isn't going to work ... that is, if the "fix" was to put that distribution check in the `while` condition of the main loop (there's been some code drift since the comment, but that makes the most sense). 

**tl;dr:** distributions are deactivated in two places, (1) when catching the `StopIteration` exception in `iterate`, and (2) when activating a new stream. Therefore, when the *last* stream is activated, `distribution` will be empty but there is still data in the stream.

:thinking-face:

**Comment by cjacoby** [comment]
I was thinking maybe this could be handled in my/our mux refactor (I'll
push a PR with what I started shortly) / #84, because the way muxes are
structured might get organized a bit to fix this. But maybe that's too much
work for this release?

**Comment by ejhumphrey** [comment]
I've three updated findings / thoughts. I'm *reasonably* confident in the assessment, so, worth sharing.

1. I don't think we can fully decouple this issue from #84 (reason to follow).
2. The logical issue here (and with #84) might effect #81, so I'd call that issue dependent on this one.
3. Diagnosis: Clarifying my previous comments, `Mux.distribution_` is being set to zero under two conditions: (1) a stream produces no samples, and (2) when a stream is activated without replacement. This causes undefined (or at least undesired) behavior for the case `with_replacement=False, revive=True, prune_empty_seeds=True`. Here, the data structure `distribution_` is being overloaded... in reality, we only want to revive 0'ed out streams that produced data, but there's no way to track this.

I propose the following:
- Fix this issue first. If this diagnosis is correct, then there's an easy fix (track dead streams explicitly), which will address the test case.
- Second, refactor the Mux semantics, e.g. PoissonMux and whatnot (half of #84)
- Confirm the statistics of PoissonMux, i.e. #81.
- Add additional Mux classes (the other half of #84)

thoughts?

**Comment by cjacoby** [comment]
:+1:

**Comment by ejhumphrey** [comment]
also, wanted to revisit letting `prune_empty_streams` potentially be False ... IIRC, I think @bmcfee's rationale once upon a time was something like "past emptiness is not indicative of future emptiness". 

While I find that a tad dangerous, I *do* support letting people shoot themselves in the foot, with adequate warnings. So, where should we put a warning about not pruning truly empty streams?
